[qunat][pt2e] Support allow_implicit_sharing flag

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113115
* __->__ #112929

Summary:
For a Node: node1 and edge: (node1, node2), since they are observing the same
Tensor, we may want to implicitly share observers, this flag allows people to
turn off this behavior for the output of the node

See the test_allow_implicit_sharing test for use case

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_implicit_sharing

Reviewers:

Subscribers:

Tasks:

Tags:

jerryzh168 committed:
[qunat][pt2e] Support allow_implicit_sharing flag

Summary:
For a Node: node1 and edge: (node1, node2), since they are observing the same
Tensor, we may want to implicitly share observers, this flag allows people to
turn off this behavior for the output of the node

See the test_allow_implicit_sharing test for use case

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_implicit_sharing

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][pt2e] Support allow_implicit_sharing flag"


Summary:
For a Node: node1 and edge: (node1, node2), since they are observing the same
Tensor, we may want to implicitly share observers, this flag allows people to
turn off this behavior for the output of the node

See the test_allow_implicit_sharing test for use case

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_implicit_sharing

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[qunat][pt2e] Support allow_implicit_sharing flag"


Summary:
For a Node: node1 and edge: (node1, node2), since they are observing the same
Tensor, we may want to implicitly share observers, this flag allows people to
turn off this behavior for the output of the node

See the test_allow_implicit_sharing test for use case

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_implicit_sharing

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1060,6 +1060,83 @@ def validate(self, model: torch.fx.GraphModule) -> None:
+               x3 -> add2 /
+               x4 -/
+
+        all add has shared input and output, and second input is using shared quantization spec pointint
```
```suggestion
        all add has shared input and output, and second input is using shared quantization spec pointing
```

kimishpatel commented on test/quantization/pt2e/test_quantize_pt2e.py:
```diff
@@ -1060,6 +1060,83 @@ def validate(self, model: torch.fx.GraphModule) -> None:
+               x4 -/
+
+        all add has shared input and output, and second input is using shared quantization spec pointint
+        to first input, but we set allow_implicit_sharing to all add nodes so input and output of add1,
```
you mean set allow_implicit_sharing to false?

jerryzh168 replied:
yeah, will fix

kimishpatel reviewed (approved):
Left some comments. looks good.

jerryzh168 committed:
Update on "[qunat][pt2e] Support allow_implicit_sharing flag"


Summary:
For a Node: node1 and edge: (node1, node2), since they are observing the same
Tensor, we may want to implicitly share observers, this flag allows people to
turn off this behavior for the output of the node

See the test_allow_implicit_sharing test for use case

Test Plan:
python test/test_quantization.py TestQuantizePT2E.test_allow_implicit_sharing

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge
